apps/gameserver/src/Player.cpp

```diff
@@ -811,6 +833,115 @@ void Player::OnItemPick(packet& p)
                 });
 }
 
+
+void Player::OnSkillUpgrade(bango::network::packet& p)
+{
+    auto index = p.pop<std::uint8_t>();
+
+    if(!m_skills.Exists(index))
+    {
+        spdlog::warn("Player {} tried to upgrade skill {} without learning it first", GetName(), index);
+        return;
+    }
+    LearnOrUpgradeSkill(index);
+}
+void Player::OnSkillLearn(bango::network::packet& p)
+{
+    const auto index = p.pop<std::uint8_t>();
+    LearnOrUpgradeSkill(index);
+}
+
+void Player::LearnOrUpgradeSkill(const std::uint8_t skill_index)
+{
+    if(!CanLearnSkill(skill_index))
+        return;
+
+    const auto* init = InitSkill::FindPlayerSkill((PLAYER_CLASS)GetClass(), skill_index);
+
+
+    auto* skill = m_skills.GetByIndex(skill_index);
+    const std::uint8_t level = skill ? skill->GetLevel() + 1 : 1;
+
+    if(level == 1) // Learn New Skill
+    {
+        auto success = m_skills.Learn(init, skill_index);
+
+        if(!success)
+        {
+            spdlog::warn("Something went wrong when Player {} tried to learn skill of id {}", GetName(), skill_index);
+            return;
+        }
+        Socket::DBClient().write(S2D_LEARNSKILL, "dbw", GetPID(), skill_index, --m_data.SUPoint);
+        spdlog::info("Player {} learned skill id {}", GetName(), skill_index);
+    }
+    else                 // Upgrade Skill
+    {
+        bool success = m_skills.Upgrade(skill_index, level);
+
+        if(!success)
+        {
+            spdlog::warn("Something went wrong when Player {} tried to upgrade skill of id {} to grade {}", GetName(), skill_index, level);
+            return;
+        }
+        Socket::DBClient().write(S2D_SKILLUP, "dbbw", GetPID(), skill_index, level, --m_data.SUPoint);
+        spdlog::info("Player {} upgraded skill id {} to level {}", GetName(), skill_index, level);
+
+    }
+    write(S2C_SKILLUP, "bb", skill_index, level);
+    SendProperty(P_SUPOINT);
+}
+
+bool Player::CanLearnSkill(const std::uint8_t index) const
+{
+    const auto* init = InitSkill::FindPlayerSkill((PLAYER_CLASS)GetClass(), index);
+
+    if(!init)
+    {
+        spdlog::warn("Player {} tried to learn an invalid skill id {}", GetName(), index);
+        return false;
+    }
+
+    if(init->LevelLimit > GetLevel())
+    {
+        spdlog::warn("Player {} tried to learn a skill id {} but is not high enough level", GetName(), index);
```

Comment:
```suggestion
        spdlog::warn("Player {} tried to learn a skill index {} but is not high enough level", GetName(), index);
```
Please re-check manually all variables and logs for skill_index skill_level